src/stirling/source_connectors/socket_tracer/bcc_bpf/socket_trace.c

```diff
@@ -151,6 +162,29 @@ static __inline void set_conn_as_ssl(uint32_t tgid, int32_t fd) {
   conn_info->ssl = true;
 }
 
+// Writes the syscall fd to a BPF map to be later consumed by the tls tracing probes in
+// addition to ensuring that set_conn_as_ssl is called. The majority of probes should call
+// this on function entry, however, there are certain probes which are not exlusively used
+// for networking and therefore must defer this fd propagation to its ret probe (where it
+// can be validated it is socket related first via sock_event).
+static __inline void propagate_fd_to_user_space_call(uint64_t pid_tgid, int fd) {
+  struct nested_syscall_fd_t* nested_syscall_fd_ptr = ssl_user_space_call_map.lookup(&pid_tgid);
+  if (nested_syscall_fd_ptr != NULL) {
+    int expected_fd = nested_syscall_fd_ptr->fd;
+    if (expected_fd != kInvalidFD && expected_fd != fd) {
+      nested_syscall_fd_ptr->mismatched_fds = true;
+    }
```

Comment:
Would this be easier to follow:
```
int current_fd = nested_syscall_fd_ptr->fd;
if (current_fd == kInvalidFD) {
  nested_syscall_fd_ptr->fd = fd;
}
else if (current_fd != fd) {
  // Found two different fds in nested syscalls, need to record this fact.
  nested_syscall_fd_ptr->mismatched_fds = true;
}
```
Maybe the existing functionality (setting the struct value on every invocation) is required.

Mostly, just trying to puzzle through what the "normal" path through this is.

Succinctly, if there is a way to write this that makes the "normal" path obvious, great. If not, perhaps we can add a comment that says "we expect to find the struct fd as invalid, and just set it. But if we see more than one nested syscall, we expect the same fd. If, however, we see more than one nested syscall and different fds, then we flag the issue."

Reply:
When I was developing this I was worried that conditionally setting the file descriptor could be a source of error. It felt safer to always have the BPF map updated to the most recent fd provided to syscall.

Reply:
Is there a good reason to always keep the most recent fd? I do see the intuition and wonder if we can pin down a solid reason. Gently suggesting a comment to explain the common case path through the code in any case.

Reply:
Here's the difference I see between the two methods:
- Existing way could detect the number of times a socket fd is mismatched (although it would require more changes to truly measure that).
- Proposed way would only detect a change from whatever was initially recorded.

I'm unfortunately at a loss for which is "more correct". This validation attempts to detect a problem that we believe is not possible and as a result I don't have a strong rationale for which one to proceed with.

I think either case will catch the situation we are trying to defend against. I went ahead and updated this to match your suggestion and included a comment explaining why we must detect this.